### PR TITLE
Pass compiler flags from driver to AST level

### PR DESCRIPTION
**Pull request by lucteo**
Previously, we used one instance of `CompilerInfo` which would not allow us to pass compiler features from driver to the AST level. This change drops the single instance and passes down a `CompilerInfo` object from driver to AST. It also adds a `freestanding` option to the driver that can now be used with `#if feature(freestanding)`

### CODE REVIEW

File: Sources/Core/AST/AST.swift [hunk 1]
```diff
@@ -50,8 +58,15 @@ public struct AST {
     set { storage.coreLibrary = newValue }
   }
 
+  /// Information about the compiler processing `self`.
+  public var compilerInfo: CompilerInfo {
```
Comment (kyouko-taiga) [inline Sources/Core/AST/AST.swift:62]:
```suggestion
  public var compiler: CompilerInfo {
```
There's no need to repeat the type of the property verbatim.

File: Sources/Core/CompilerInfo.swift [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /// Describes the compiler; used as source of truth for conditional compilation.
-public struct CompilerInfo {
+public struct CompilerInfo: Codable {
```
Comment (kyouko-taiga) [inline Sources/Core/CompilerInfo.swift:2]: Can we avoid abbreviations?
```suggestion
public struct CompilerConfiguration: Codable {
```

**Review by kyouko-taiga — COMMENTED** [review]
I left a couple of suggestions inline, but it already looks pretty good.

File: Sources/Core/AST/AST.swift (hunk 1 above)
Comment (lucteo) [reply Sources/Core/AST/AST.swift]: fixed in all places.

File: Sources/Core/CompilerInfo.swift (hunk 2 above)
Comment (lucteo) [reply Sources/Core/CompilerInfo.swift]: fixed in all places.

**Pull request merged**